### PR TITLE
New: Add offline support using service workers (fixes #23)

### LINKED ISSUE #23: Make it work offline
Ideas or PR welcome.

### PR DESCRIPTION
**Pull request by gyandeeps**
Currently, NextJS doesn't have built-in support for service worker/offline. ref https://github.com/zeit/next.js/issues/861

Since we do not have a server and we only use static pages, we have to copy the generated service-worker.js file to `out` folder.
I did test on my local machine and it works when u run a [temporary server](https://stackoverflow.com/a/29046869/2449389) from out directory to feed the files.

### CODE REVIEW

File: .npmrc [hunk 1]
```diff
@@ -0,0 +1 @@
+package-lock=false
```
Comment (ritz078) [inline .npmrc]: No need to do this. It is fine if you commit the lockfile.

**Comment by ritz078** [comment]
This project gets deployed a lot. So will have to test this for cache bursting. Also we can prompt the user about changes (So that he can reload if there is any update). What do you think. This PR needs some testing so that there is no issue with the cache. I will deploy this and give you the temporary URL.

**Comment by ritz078** [comment]
It is deployed [here](https://out-sgqjotvhkg.now.sh). You can create a free account on zeit.co and deploy this to a server using [now-cli](https://github.com/zeit/now-cli)

```
cd transform-www
yarn deploy
```

It seems this PR is dependent on the PWA one ? As manifest is needed. You can merge both into one so that its easier to debug.

**Comment by gyandeeps** [comment]
> It seems this PR is dependent on the PWA one ?

Not exactly. But they do work together nicely..

> We need a robust cache bursting technique too. And a prompt to show the user that there is an update available.

Its network first, so user will get the updates.

File: .npmrc (hunk 1 above)
Comment (gyandeeps) [reply .npmrc]: I mean we will end up having .yarn.lock and package-lock.json, are you ok with that?
Then some contributors will update one vs the other...

Comment (ritz078) [reply .npmrc]: Yes that may be an issue. Let it be in gitignore

File: pages/index.js [hunk 2]
```diff
@@ -30,6 +30,19 @@ export default class Main extends PureComponent {
     ${code}`;
   };
 
+  async componentDidMount() {
```
Comment (ritz078) [inline pages/index.js]: why do you need this to be async ?

File: .npmrc (hunk 1 above)
Comment (gyandeeps) [reply .npmrc]: adding it in gitignore creates some issue when you try to release stuff (I have seen some issues on other projects)... this is the best way to manage this...

**Comment by gyandeeps** [comment]
Demo: https://out-nskzusyzog.now.sh/

**Comment by gyandeeps** [comment]
I might need some help in images as I am so bad in that.

**Comment by ritz078** [comment]
I will do that part. Have you tried going to url like `/css-to-js` and then refreshing it in offline mode ?

**Comment by gyandeeps** [comment]
It works... you can try the demo page: https://transform-pwa.now.sh/
As for the current setup, a person has to visit that page before it can be come offline compatible....

**Comment by ritz078** [comment]
Yes it does. my bad 😞  It was a problem with opened devtools where the service worker was inactivated. 

We can actually cache all the pages. Next JS prefetches all the links that are available in the DOM. Unfortunately in the navigation accordion only the group which is opened has the rendered children which has all the `Link`. So the pages from other groups are not pre-fetched since the `Link` for them wasn't rendered at that time.

**Comment by ritz078** [comment]
I have found the solution to that. Programatically fetching URL fixes that. Can you give me access to your PR by checking `Allow edits from maintainers.` on you PR so that I can make some changes and update the images ?

**Comment by gyandeeps** [comment]
you have permission already... thanks

**Comment by ritz078** [comment]
Here is an issue. (Do this in incognito)

- Open A
- Navigate to B
- Offline 
- Refresh (It breaks)

Because the request is being made to the server `/B` which was never made earlier as only the JS was prefetched. But since we already have the JS available with us (cached), maybe there can be some solution to that.

On the other hand

- Open A
- Navigate to B
- Offline
- Open A
- Goto any route and it works great.

So even though we have all the files available on the client side we are not able to use those in first case.

**Comment by ritz078** [comment]
It was an awesome PR ❤️ . Thanks a lot for your contribution. 🎉

**Comment by ritz078** [comment]
It is live now.

**Pull request merged**